File: web_client/data/mappers/ListingMapper.ts
import { ListingDTO, ListingPayloadDTO } from '../models/ListingDTO';
import { Listing } from '../../domain/models/Listing';
import { ListingPayload } from '../../domain/models/ListingPayload';

export class ListingMapper {
  static toDomain(dto: ListingDTO): Listing {
    return {
      id: dto.id,
      title: dto.title,
      description: dto.description,
      priceAmount: dto.price_amount,
      priceCurrency: dto.price_currency,
      isPriceNegotiable: dto.is_price_negotiable,
      condition: dto.condition,
      dealType: dto.deal_type,
      sellerType: dto.seller_type,
      categoryId: dto.category || 0,
      categoryName: dto.category_name,
      locationId: dto.location || 0,
      locationName: dto.location_name,
      lat: dto.lat,
      lon: dto.lon,
      mediaUrls: dto.media_urls,
      attributes: dto.attributes,
      status: dto.status,
      createdAt: dto.created_at,
      updatedAt: dto.updated_at,
      refreshedAt: dto.refreshed_at,
      user: dto.user,
      seller: dto.seller,
    };
  }

  static toDomainList(dtos: ListingDTO[]): Listing[] {
    return dtos.map(dto => this.toDomain(dto));
  }

  static payloadToDTO(payload: ListingPayload): ListingPayloadDTO {
    return {
      title: payload.title,
      description: payload.description,
      price_amount: payload.priceAmount,
      price_currency: payload.priceCurrency,
      is_price_negotiable: payload.isPriceNegotiable,
      condition: payload.condition,
      deal_type: payload.dealType,
      seller_type: payload.sellerType,
      category: payload.categoryId,
      location: payload.locationId,
      lat: payload.lat,
      lon: payload.lon,
      attributes: payload.attributes?.map(attr => ({
        attribute: attr.attributeId,
        value: attr.value,
      })),
    };
  }

  static partialPayloadToDTO(payload: Partial<ListingPayload>): Partial<ListingPayloadDTO> {
    const dto: Partial<ListingPayloadDTO> = {};

    if (payload.title !== undefined) dto.title = payload.title;
    if (payload.description !== undefined) dto.description = payload.description;
    if (payload.priceAmount !== undefined) dto.price_amount = payload.priceAmount;
    if (payload.priceCurrency !== undefined) dto.price_currency = payload.priceCurrency;
    if (payload.isPriceNegotiable !== undefined) dto.is_price_negotiable = payload.isPriceNegotiable;
    if (payload.condition !== undefined) dto.condition = payload.condition;
    if (payload.dealType !== undefined) dto.deal_type = payload.dealType;
    if (payload.sellerType !== undefined) dto.seller_type = payload.sellerType;
    if (payload.categoryId !== undefined) dto.category = payload.categoryId;
    if (payload.locationId !== undefined) dto.location = payload.locationId;
    if (payload.lat !== undefined) dto.lat = payload.lat;
    if (payload.lon !== undefined) dto.lon = payload.lon;
    if (payload.attributes !== undefined) {
      dto.attributes = payload.attributes?.map(attr => ({
        attribute: attr.attributeId,
        value: attr.value,
      }));
    }

    return dto;
  }
}

File: mobile_client/lib/data/mappers/listing_mapper.dart
/// Listing mappers for ListingDto ↔ Listing conversion.
library;

import '../../domain/entities/listing.dart';
import '../../domain/entities/user.dart';
import '../models/listing/listing_dto.dart';

/// Mapper for ListingMediaDto → ListingMedia.
extension ListingMediaDtoMapper on ListingMediaDto {
  ListingMedia toEntity() {
    return ListingMedia(
      id: id,
      type: type,
      image: image,
      imageUrl: imageUrl,
      width: width,
      height: height,
      order: order,
      uploadedAt: uploadedAt,
    );
  }
}

/// Mapper for ListingAttributeDto → ListingAttribute.
extension ListingAttributeDtoMapper on ListingAttributeDto {
  ListingAttribute toEntity() {
    return ListingAttribute(
      attributeId: attribute,
      key: key,
      value: value,
      label: label,
      type: type,
    );
  }
}

/// Mapper for ListingUserDto → ListingUser.
extension ListingUserDtoMapper on ListingUserDto {
  ListingUser toEntity() {
    return ListingUser(
      id: id,
      phone: phone,
      name: name,
      displayName: displayName,
      phoneE164: phoneE164,
    );
  }
}

/// Mapper for ListingSellerDto → ListingSeller.
extension ListingSellerDtoMapper on ListingSellerDto {
  ListingSeller toEntity() {
    return ListingSeller(
      id: id,
      name: name,
      avatarUrl: avatarUrl,
      since: since,
      logo: logo,
      banner: banner,
      phone: phone,
      lastActiveAt: lastActiveAt,
    );
  }
}

/// Mapper for ListingDto → Listing.
extension ListingDtoMapper on ListingDto {
  Listing toEntity() {
    return Listing(
      id: id,
      title: title,
      description: description,
      priceAmount: priceAmount,
      priceCurrency: priceCurrency,
      isPriceNegotiable: isPriceNegotiable,
      condition: condition,
      dealType: dealType,
      sellerType: sellerType,
      categoryId: category ?? 0,
      categoryName: categoryName,
      categorySlug: categorySlug,
      locationId: location ?? 0,
      locationName: locationName,
      locationSlug: locationSlug,
      lat: lat,
      lon: lon,
      media: media?.map((m) => m.toEntity()).toList(),
      mediaUrls: mediaUrls,
      attributes: attributes?.map((a) => a.toEntity()).toList(),
      status: status,
      contactName: contactName,
      contactEmail: contactEmail,
      contactPhone: contactPhone,
      createdAt: createdAt != null ? DateTime.tryParse(createdAt!) : null,
      updatedAt: updatedAt != null ? DateTime.tryParse(updatedAt!) : null,
      refreshedAt: refreshedAt != null ? DateTime.tryParse(refreshedAt!) : null,
      expiresAt: expiresAt != null ? DateTime.tryParse(expiresAt!) : null,
      qualityScore: qualityScore,
      contactPhoneMasked: contactPhoneMasked,
      priceNormalized: priceNormalized,
      isPromoted: isPromoted,
      viewCount: viewCount ?? 0,
      favoriteCount: favoriteCount ?? 0,
      interestCount: interestCount ?? 0,
      userId: userId,
      user: user?.toEntity(),
      seller: seller?.toEntity(),
    );
  }
}

/// Mapper for ListingPayload → ListingPayloadDto.
extension ListingPayloadMapper on ListingPayload {
  ListingPayloadDto toDto() {
    return ListingPayloadDto(
      title: title,
      description: description,
      priceAmount: priceAmount,
      priceCurrency: priceCurrency,
      isPriceNegotiable: isPriceNegotiable,
      condition: condition,
      dealType: dealType,
      sellerType: sellerType,
      category: categoryId,
      location: locationId,
      lat: lat,
      lon: lon,
      attributes: attributes?.map((a) => a.toDto()).toList(),
      contactName: contactName,
      contactEmail: contactEmail,
      contactPhone: contactPhone,
    );
  }
}

/// Mapper for ListingAttributePayload → ListingAttributePayloadDto.
extension ListingAttributePayloadMapper on ListingAttributePayload {
  ListingAttributePayloadDto toDto() {
    return ListingAttributePayloadDto(
      attribute: attributeId,
      value: value,
    );
  }
}
